File: Math and Geometry/Python/Maximum Swap.py
# Problem link: https://leetcode.com/problems/maximum-swap/description/
# Time: O(n log n)
# Space: O(n)
class Solution:
    def maximumSwap(self, num: int) -> int:
        store, temp, result = [], num, 0

        while temp:
            store.insert(0, temp % 10)
            temp //= 10
        
        temp = sorted(store, reverse = True)

        for i in range(len(temp)):
            if temp[i] != store[i]:
                idx = len(store) - 1

                while store[idx] != temp[i]:
                    idx -= 1
                
                store[i], store[idx] = store[idx], store[i]
                break

        for n in store:
            result = result * 10 + n
        
        return result    

File: Math and Geometry/Java/Maximum Swap.java
// Problem link: https://leetcode.com/problems/maximum-swap/description/
// Time: O(n log n)
// Space: O(n)
class Solution {
    public int maximumSwap(int num) {
        List<Integer> store = new ArrayList<>();
        int temp = num, result = 0, idx;

        while(temp > 0) {
            store.add(0, temp % 10);
            temp /= 10;
        }

        List<Integer> sortedList = new ArrayList<>(store);
        sortedList.sort(Collections.reverseOrder());

        for(int i = 0; i < store.size(); i++) {
            if(sortedList.get(i) != store.get(i)) {
                idx = store.size() - 1;

                while(store.get(idx) != sortedList.get(i)) {
                    idx--;
                }
                
                temp = store.get(i);
                store.set(i, store.get(idx));
                store.set(idx, temp);
                break;
            }
        }


        for(int n : store) {
            result = result * 10 + n;
        }

        return result;
    }
}
